tests/test_reweighting.py

```diff
@@ -0,0 +1,407 @@
+
+
+def make_gaussian_testsystem():
+    """normalized/unnormalized 1D Gaussian with a dependence on lambda and params"""
+
+    def annealed_gaussian_def(lam, params):
+        initial_mean, initial_log_sigma = 0.0, 0.0
+        target_mean, target_log_sigma = params
+
+        # lam = 0 -> (mean = 0, stddev = 1)
+        # lam = 1 -> (mean = target_mean, stddev = target_sigma)
+        mean = lam * target_mean - (1 - lam) * initial_mean
+        stddev = np.exp(lam * target_log_sigma + (1 - lam) * initial_log_sigma)
+
+        return mean, stddev
+
+    def sample(lam, params, n_samples):
+        mean, stddev = annealed_gaussian_def(lam, params)
+        return norm.rvs(loc=mean, scale=stddev, size=(n_samples, 1))
+
+    def logpdf(x, lam, params):
+        mean, stddev = annealed_gaussian_def(lam, params)
+        return np.sum(norm_logpdf(x, loc=mean, scale=stddev))
+
+    def u_fxn(x, lam, params):
+        """unnormalized version of -logpdf"""
+        mean, stddev = annealed_gaussian_def(lam, params)
+        return np.sum(0.5 * ((x - mean) / stddev) ** 2)
+
+    def normalized_u_fxn(x, lam, params):
+        return -logpdf(x, lam, params)
+
+    def reduced_free_energy(lam, params):
+        mean, stddev = annealed_gaussian_def(lam, params)
+        log_z = np.log(stddev * np.sqrt(2 * np.pi))
+        return -log_z
+
+    return u_fxn, normalized_u_fxn, sample, reduced_free_energy
+
+
+def assert_estimator_accurate(estimate_delta_f, analytical_delta_f, ref_params, n_random_trials, atol):
+    """for many random parameter sets, assert that the reweighted estimates of
+    delta_f(params) and grad(delta_f)(params) are consistent with analytical_delta_f function
+
+    Parameters
+    ----------
+    estimate_delta_f: Callable[[params], float]
+        estimated free energy difference as a
+        differentiable function of params
+    analytical_delta_f: Callable[[params], float]
+        exact free energy difference as a
+        differentiable function of params
+    ref_params: [2,] array
+        generate random parameters near ref_params
+    n_random_trials: int
+    atol: float
+        absolute tolerance
+    """
+
+    def sample_random_params():
+        mean = ref_params[0] + onp.random.rand()
+        log_sigma = ref_params[1] - onp.random.rand()
+        return np.array([mean, log_sigma])
+
+    f_hat, g_hat = value_and_grad(estimate_delta_f)(ref_params)
+    f_ref, g_ref = value_and_grad(analytical_delta_f)(ref_params)
+
+    onp.testing.assert_allclose(f_hat, f_ref, atol=atol)
+    onp.testing.assert_allclose(g_hat, g_ref, atol=atol)
+
+    for _ in range(n_random_trials):
+        trial_params = sample_random_params()
+        f_hat, g_hat = value_and_grad(estimate_delta_f)(trial_params)
+        f_ref, g_ref = value_and_grad(analytical_delta_f)(trial_params)
+
+        onp.testing.assert_allclose(f_hat, f_ref, atol=atol)
+        onp.testing.assert_allclose(g_hat, g_ref, atol=atol)
```

Comment:
How should we think about precision of `dDeltaF/dParams` both for the gaussian case, and in general?  For the gaussian case, I'd imagine that `dDeltaF/dlog_sigma` might require more precision than `dDeltaF/dmean`. Presumably, an acceptable level of error is proportional to the internal learning rate one might take when optimizing the parameters, so maybe `rtol` is more approriate here?

Reply:
Will need to think about this.

Reply:
>For the gaussian case, I'd imagine that dDeltaF/dlog_sigma might require more precision than dDeltaF/dmean. 

I'd imagine so.

In a [documentation notebook](https://github.com/proteneer/timemachine/blob/c39f6dc765ffc8e7dfcdc1920daff02ca2df962e/examples/free_energy_gradients_comparison/comparison_on_1d_system.ipynb) of #404 , cell [46], note that estimates of the different gradient components have different variance.

Reply:
>How should we think about precision of dDeltaF/dParams both for the gaussian case, and in general?

Preferred interpretation: Exact gradient of a randomized approximation of `DeltaF(params)`.

`delta_f_approx_fxn = construct_approximation(ref_params, random_seed)`

>Presumably, an acceptable level of error is proportional to the internal learning rate one might take when optimizing the parameters

In the case that you created a fresh random approximation every optimization step

```python
delta_f_approx_fxn_0 = construct_approximation(params_0, seed_0)

params_1 = params_0 + stepsize * grad(delta_f_approx_fxn_0)(params_0)

delta_f_approx_fxn_1 = construct_approximation(params_1, seed_1)

params_2 = params_1 + stepsize * grad(delta_f_approx_fxn_1)(params_1)
...
```

then appropriate step size would be a function of both the geometry of `delta_f_exact_fxn` and the precision of the gradient estimate.

Instead, I think we would want to perform iterates like

```python
delta_f_approx_fxn_0 = construct_approximation(params_0, seed_0)

params_1 = local_optimize(fun=delta_f_approx_fxn_0, x0=params_0, ...)

delta_f_approx_fxn_1 = construct_approximation(params_1, seed_1)

params_2 = local_optimize(fun=delta_f_approx_fxn_1, x0=params_1, ...)
...
```

where `local_optimize` returns the local optimum (+ some regularization + optional restriction to some trust region, glossing over how that trust region is defined.)

Then whatever step sizes etc. are used inside `local_optimize` do NOT depend on the precision of the estimate, only on "the geometry of the problem."

The amount of progress you make in each outer loop step (construct_approximation, call local_optimize) would be influenced by how that trust region is defined.

(There's a similar way to view gradient descent: each step of SGD is exactly optimizing a random local approximation of f + some regularization.)

Reply:
Sorry, missed replying to this part:

>so maybe rtol is more approriate here?

I don't think so.

`isclose(estimated_gradient, exact_gradient, rtol=...)` might be appropriate if the noise magnitude of each component `estimated_gradient[i]` was proportional to the magnitude of the corresponding gradient component `abs(exact_gradient[i])`.

(Counterexample: [1D Lennard-Jones testsystem](https://github.com/proteneer/timemachine/blob/c39f6dc765ffc8e7dfcdc1920daff02ca2df962e/examples/free_energy_gradients_comparison/comparison_on_1d_system.ipynb): exact gradient w.r.t. sigma, epsilon parameters ~= [-0.15, -0.3] (cell [43]), but estimates of these gradient components have marginal variance ~= [0.1, 0.01] (cell [41]).)

Reply:
I'm going to leave this "unresolved" so I can find the comments above easier. But this mostly addresses my concerns.

Reply:
Opened https://github.com/proteneer/timemachine/issues/667 for continued discussion after this PR